File: server/apps/nexotype/schemas/engineering/candidate_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class CandidateCreate(BaseModel):
    """Schema for creating a new candidate"""
    parent_candidate_id: int | None = None
    asset_id: int
    version_number: str


class CandidateUpdate(BaseModel):
    """Schema for updating a candidate"""
    parent_candidate_id: int | None = None
    asset_id: int = None
    version_number: str = None


# ==========================================
# Response Schemas
# ==========================================

class CandidateDetail(BaseModel):
    """Schema for candidate details"""
    id: int
    parent_candidate_id: int | None = None
    asset_id: int
    version_number: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class CandidateListResponse(BaseModel):
    """Response schema for listing candidates"""
    success: bool
    data: list[CandidateDetail] | None = None
    count: int = 0
    error: str | None = None


class CandidateResponse(BaseModel):
    """Response schema for single candidate operations"""
    success: bool
    data: CandidateDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/engineering/candidate.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const CandidateSchema = z.object({
  id: z.number(),
  parent_candidate_id: z.number().int().nullable().optional(),
  asset_id: z.number().int(),
  version_number: z.string().min(1).max(20),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateCandidateSchema = z.object({
  parent_candidate_id: z.number().int().nullable().optional(),
  asset_id: z.number().int(),
  version_number: z.string().min(1).max(20),
});

export const UpdateCandidateSchema = z.object({
  parent_candidate_id: z.number().int().nullable().optional(),
  asset_id: z.number().int().optional(),
  version_number: z.string().min(1).max(20).optional(),
});

// ==========================================
// Types
// ==========================================

export type Candidate = z.infer<typeof CandidateSchema>;
export type CreateCandidate = z.infer<typeof CreateCandidateSchema>;
export type UpdateCandidate = z.infer<typeof UpdateCandidateSchema>;

// ==========================================
// Response Types
// ==========================================

export type CandidateResponse = {
  success: boolean;
  data?: Candidate;
  error?: string;
};

export type CandidatesResponse = {
  success: boolean;
  data?: Candidate[];
  count?: number;
  error?: string;
};
